substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/util/LLVMDebugLineInfo.java

```diff
@@ -0,0 +1,205 @@
+            } else if (javaMethod instanceof CustomSubstitutionMethod) {
+                return ((CustomSubstitutionMethod) javaMethod).getOriginal().getDeclaringClass();
+            }
+            return hostedType.getWrapped().getWrapped();
+        }
+        ResolvedJavaMethod javaMethod = hostedMethod.getWrapped().getWrapped();
+        return javaMethod.getDeclaringClass();
+    }
+
+    public String fileName() {
+        if (fullFilePath != null) {
+            Path fileName = fullFilePath.getFileName();
+            if (fileName != null) {
+                return fileName.toString();
+            }
+        }
+        return null;
+    }
+
+    public Path filePath() {
+        if (fullFilePath != null) {
+            return fullFilePath.getParent();
+        }
+        return null;
+    }
+
+    public Path cachePath() {
+        return cachePath;
+    }
+
+    public String className() {
+        return method.format("%H");
+    }
+
+    public String methodName() {
+        ResolvedJavaMethod targetMethod = method;
+        while (targetMethod instanceof WrappedJavaMethod) {
+            targetMethod = ((WrappedJavaMethod) targetMethod).getWrapped();
+        }
+        if (targetMethod instanceof SubstitutionMethod) {
+            targetMethod = ((SubstitutionMethod) targetMethod).getOriginal();
+        } else if (targetMethod instanceof CustomSubstitutionMethod) {
+            targetMethod = ((CustomSubstitutionMethod) targetMethod).getOriginal();
+        }
+        String name = targetMethod.getName();
+        if (name.equals("<init>")) {
+            if (method instanceof HostedMethod) {
+                name = getJavaType((HostedMethod) method, true).toJavaName();
+                if (name.indexOf('.') >= 0) {
+                    name = name.substring(name.lastIndexOf('.') + 1);
+                }
+                if (name.indexOf('$') >= 0) {
+                    name = name.substring(name.lastIndexOf('$') + 1);
+                }
+            } else {
+                name = targetMethod.format("%h");
+                if (name.indexOf('$') >= 0) {
+                    name = name.substring(name.lastIndexOf('$') + 1);
+                }
+            }
+        }
+        return name;
+    }
+
+    public String symbolNameForMethod() {
+        return NativeImage.localSymbolNameForMethod(method);
+    }
+
+    public int addressLo() {
+        throw unimplemented();
+    }
+
+    public int addressHi() {
+        throw unimplemented();
+    }
+
+    public int line() {
+        LineNumberTable lineNumberTable = method.getLineNumberTable();
+        if (lineNumberTable != null) {
+            return lineNumberTable.getLineNumber(bci);
+        }
+        return 0;
+    }
+
+    @SuppressWarnings("try")
+    public void computeFullFilePath() {
+        ResolvedJavaType declaringClass = method.getDeclaringClass();
+        Class<?> clazz = null;
+        if (declaringClass instanceof OriginalClassProvider) {
```

Comment:
Since you're generating debug info during compilation, `method` should always be a `HostedMerthod`. This should allow you to use `getJavaClass()` directly here to get the class information.

Reply:
That is true of classes that are presented as the declaring class of a compiled method. It did not turn out to be true of all classes that turned up as declaring classes of some inlined methods that are referenced in frame data derived from the CompilationResult. These inline classes also need to be associated with a source file. 

Note also that in this case the source file for a substituted method S.m1 needs to be the file for class S, not for the class C that it substitutes. That is why the lookup functions above take a flag describing whether or not the original is wanted.

Reply:
That makes sense, I didn't consider inlined methods. As a general question related to that file, how is it different for LLVM? Couldn't we just reuse the one from the normal backend?